### PR TITLE
*: use context to manage RaftCluster

### LINKED ISSUE #4293: Only use context to manage RaftCluster
## Development Task

Currently, there are both `ctx context.Context` and `quit chan struct{}` in `RaftCluster` to control the start and stop, where `ctx` is actually `Server.ctx`. Currently, it has the following problems:

* The control sub-module needs to pass parameters `ctx` and `quit` at the same time
* The `suspectRegions` and `suspectKeyRanges` only use `ctx` without `quit`, which leads to leaks after changing from Leader to Follower.
* `RegionSyncer` did not make full use of quit to terminate the service as soon as possible.

We can solve them by combining `ctx` and `quit`.

### PR DESCRIPTION
**Pull request by HunDunDM**
### What problem does this PR solve?

close #4293

### What is changed and how it works?

* `RaftCluster` uses its own `ctx` to replace the original `quit`.
* `RegionSyncer` can stop immediately when `ctx` is `Done` during service.

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
None
```

### CODE REVIEW

File: server/region_syncer/server.go [hunk 1]
```diff
@@ -42,6 +42,11 @@ const (
 	defaultHistoryBufferSize = 10000
 )
 
+var (
+	// ErrNotStarted is returned when current server is starting.
+	ErrNotStarted = status.Errorf(codes.Unavailable, "server not started")
```
Comment (rleungx) [inline server/region_syncer/server.go:47]: How about reducing the duplicated definision?

File: server/region_syncer/server.go [hunk 2]
```diff
@@ -148,7 +166,21 @@ func (s *RegionSyncer) RunServer(regionNotifier <-chan *core.RegionInfo, quit ch
 // Sync firstly tries to sync the history records to client.
 // then to sync the latest records.
 func (s *RegionSyncer) Sync(stream pdpb.PD_SyncRegionsServer) error {
+	s.mu.RLock()
+	ctx := s.mu.serverCtx
+	s.mu.RUnlock()
+
+	if ctx == nil {
```
Comment (rleungx) [inline server/region_syncer/server.go:173]: When will it become nil?

Comment (HunDunDM) [reply server/region_syncer/server.go:173]: `RaftCluster` is not started.

File: server/region_syncer/server.go (hunk 1 above)
Comment (HunDunDM) [reply server/region_syncer/server.go:47]: I didn't think of a good way. Maybe only a normal error is returned in region_syncer?

File: server/region_syncer/server.go (hunk 2 above)
Comment (rleungx) [reply server/region_syncer/server.go:173]: Why not move this check to server so that we can avoid the duplicated definition.

**Review by rleungx — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**